### PR TITLE
Handle IntroductionPoint death

### PR DESCRIPTION
**Pull request by parhamsaremi**
Kill TorServiceHost and inform it about the death of IntroductionPoints. Currently, TorServiceHost doesn't care about IntroductionPoints' deaths, and continues even when it shouldn't.

### CODE REVIEW

**Comment by knocte** [comment]
3 nits:
1. Please put a description for PR (you can copy+paste from one of the commit messages, the most important one; or do a summary of both).
2. Please state the reason why this can't be merged yet.
3. You have a typo in your commit: pionts -> points.

**Comment by knocte** [comment]
> Handle introductionpoint

I believe you meant `IntroductionPoint` in the title of the PR.

**Comment by knocte** [comment]
When overwriting the branch of this PR, shouldn't you guys say if it's ready to review/merge or why it's not? @aarani @parhamsaremi

**Comment by aarani** [comment]
Yes this should be ready to review

File: NOnion/Network/TorCircuit.fs [hunk 1]
```diff
@@ -411,16 +449,19 @@ and TorCircuit
                         fromNode.Window.PackageIncrease()
                     | RelayIntroduce2 introduceMsg ->
                         match circuitState with
-                        | ReadyAsIntroductionPoint(_, _, _, _, callback) ->
-                            do! callback(introduceMsg)
+                        | ReadyAsIntroductionPoint
+                            (
+                                _, _, _, _, callback, _disconnectionCallback
+                            ) -> do! callback(introduceMsg)
```
Comment (knocte) [inline NOnion/Network/TorCircuit.fs:455]: @aarani no need for parens

File: NOnion/Network/TorCircuit.fs [hunk 2]
```diff
@@ -1203,10 +1265,20 @@ and TorCircuit
...
         (callback: Func<RelayIntroduce, Task>)
+        (disconnectCallback: Action)
         =
-        fun relayIntroduce ->
-            async { return! callback.Invoke(relayIntroduce) |> Async.AwaitTask }
-        |> self.RegisterAsIntroductionPoint authKeyPairOpt
+        let introduceCallback =
+            fun relayIntroduce ->
+                async {
+                    return! callback.Invoke(relayIntroduce) |> Async.AwaitTask
```
Comment (knocte) [inline NOnion/Network/TorCircuit.fs:1273]: @aarani no need for parens

File: NOnion/Services/TorServiceHost.fs [hunk 3]
```diff
@@ -766,9 +790,16 @@ type TorServiceHost
 
             let rec getConnectionRequest() =
                 async {
-                    do!
-                        newClientSemaphore.WaitAsync(cancelToken)
-                        |> Async.AwaitTask
+                    try
+                        do!
+                            newClientSemaphore.WaitAsync(linkedCts.Token)
```
Comment (knocte) [inline NOnion/Services/TorServiceHost.fs:795]: @aarani no need for parens

File: NOnion/Services/TorServiceHost.fs [hunk 4]
```diff
@@ -95,6 +101,17 @@ type TorServiceHost
...
         }
 
+    member private self.IntroductionPointDeathCallBack() =
+        lock
+            incrementLock
+            (fun () ->
+                introductionPointDeadCounter <- introductionPointDeadCounter + 1
+
+                if introductionPointDeadCounter > Constants.HiddenServices.IntroductionPointCount
+                                                  / 2 then
```
[the remark this answers is not in the document]
Comment (aarani) [reply NOnion/Services/TorServiceHost.fs:111]: Not really the point of this line is that if more than half of the IPs are dead torservicehost is also dead

[the remark this answers is not in the document]
Comment (knocte) [reply NOnion/Services/TorServiceHost.fs:111]: that definitely deserves a comment then

[the remark this answers is not in the document]
Comment (aarani) [reply NOnion/Services/TorServiceHost.fs:111]: I added a comment for this

File: NOnion/Services/TorServiceHost.fs [hunk 5]
```diff
@@ -95,6 +101,18 @@ type TorServiceHost
             queueLock.RunSyncWithSemaphore registerConnectionRequest
         }
 
+    member private self.IntroductionPointDeathCallBack() =
```
Comment (knocte) [inline NOnion/Services/TorServiceHost.fs:104]: @aarani should be Callback, not CallBack

**Pull request merged**